Faster tests

crocha700 opened the pull request:
Currently the tests are too slow #62 

```{bash}
$ time nosetests 
(...)
----------------------------------------------------------------------
Ran 26 tests in 56.840s

OK

real	0m57.094s
user	0m56.450s
sys	0m0.391s
```

This PR makes some changes in the reference solutions "tests". First, it refactors the three reference solutions into a single file. Second, it changes the reference solutions for ```sqg``` and ```bt``` to make the tests run faster. Those were the bottle necks as per profile reported in #62. 

(I kept the ```two_layer``` reference solution the same since that was the original solution... But when running this tests many times on my computer the ```two_layer``` reference solution test fails every once in a while.This issue was previously reported in #96. I figure that the problem is that the time step is too large in that test.)

With these changes:
```{bash}
$ time nosetests 
(...)
----------------------------------------------------------------------
Ran 26 tests in 12.711s

OK

real	0m13.174s
user	0m12.684s
sys	0m0.279s
```

Thus these changes will make the suit of tests to run about 5 times faster.

rabernat commented on pyqg/tests/test_twolayer_qg.py:
```diff
@@ -1,112 +0,0 @@
-        'APEgenspec': 9.058591846403974e-05,
-        'KE1spec': 3.338261440237941e+03,
-        'KE2spec': 7.043282793801889e+01
-    }
```
At some point we will want to put the diagnostics tests back in, no?

crocha700 replied:
We might, but with the currently values the inconsistency is wild:

```{python}
EKEdiss
   EKEdiss: 1.114800472490948e-10 
            6.368668363708053e-08 (desired)
APEgen
    APEgen: 1.260862629353291e-07 
            8.842056320175081e-08 (desired)
APEgenspec: 1.291480122136782e-04 
            9.058591846403974e-05 (desired)
   entspec: 8.172722867144368e-07 
            5.703438193477885e-07 (desired)
   APEflux: 1.347717697982667e-04 
            9.192940039964286e-05 (desired)
    KEflux: 2.653355719569081e-04 
            1.702621259427053e-04 (desired)
```

May be we could restart this test with a new reference solution.

MFJansen replied:
Uhm, according to this EKEdiss has changed by over two orders of magnitude. I'm sure this is significant! The new values seem wrong to me, since EKEdiss is more than two orders of magnitude smaller than EKEgen. 
 
The other differences actually seem kind of significantly different to me as well. We should probably figure out what's going on. Unfortunately this test is a mess - remember issue #2? At the same time it's the only test that goes all the way back, so we should probably figure it out.
Reading through the discussion on issue #2 again, I realize that the conversation wondered quite a bit, and ended up being mostly about bitwise reproducibility. However, it looks like it started with inconsistencies that were clearly far greater.

rabernat replied:
> it's the only test that goes all the way back

I just want to point out that, at this point, the current pyqg code has received far more scrutiny that the "original" python script. We have reproduced several classic papers and developed quite thorough tests of certain aspects of the code. There are still big gaps in the testing of course, especially regarding the diagnostics. But maintaining reproducibility of the "original" code might be a red herring at this point.

My first guess is that this order-of-magnitude issue is related to the normalization of the FFTs.

rabernat replied:
Also there are similar order-of-magnitude differences between the "old values" (commented out in the code) and the current values. Does anyone remember when that got changed?

crocha700 replied:
>My first guess is that this order-of-magnitude issue is related to the normalization of the FFTs.

Agreed. The KE spectrum is now being normalized by ```m.M^2`` (the issue of normalization is not settled yet #90). This explains the KEdiss order-of-magnitude difference.

MFJansen replied:
> I just want to point out that, at this point, the current pyqg code has received far more scrutiny than the "original" python script. We have reproduced several classic papers and developed quite thorough tests of certain aspects of the code.

I guess I'm a little less optimistic about this than you are. Let's not forget that until very recently we didn't realize that we had the sign of the advection term wrong (and in fact it was this experiment that made Cesar find this error - since he realized that the sigs of the diagnostics in the notebook had to be reversed.) 
Have we reproduced (quantitatively) classical two-layer results?
And as you are saying: I can't see any test of the diagnostics.  

Not to be a Debby Downer here, but I think we should take this seriously.

crocha700 replied:
>Have we reproduced (quantitatively) classical two-layer results?

I have reproduced Larchev & Held 1995 for a talk a gave in the Theory Seminar class that Paola is organizing this quarter. The simulation was done with the correct sign of advection. The diagnostics were computed offline because LH95 work with modal formulation.
[My_Talk_TheorySeminar_FA2015.pdf](https://github.com/pyqg/pyqg/files/26258/My_Talk_TheorySeminar_FA2015.pdf)

francispoulin replied:
Nice talk @crocha700 !

Also, I think that the SQG run reproduces the results of Held et al., thanks to @crocha700  fixing something.

rabernat replied:
> Not to be a Debby Downer here, but I think we should take this seriously.

Ok, yes, I agree. If it is not something simple like M^2, it will be quite a difficult task to track down the source of the disagreement. We will have to take the whole model apart and put it back together.

rabernat replied:
If I roll back to e7c5ee2ed204c640400517e4c820f75918666185, the test passes.

rabernat replied:
One issue is that the ``tavestart`` parameter is ignored in that early version.

```python
        # here is where we calculate diagnostics                                  
        if (self.t>=self.dt) and (self.tc%self.taveints==0):                      
            self._increment_diagnostics() 
```

If I set ``tavestart==0`` in the latest version, I get much better (but still not perfect) agreement
```
   EKEdiss: 7.502854837145852e-11 
            6.368668363708053e-08 (desired)
    APEgen: 8.571462649785877e-08 
            8.842056320175081e-08 (desired)
APEgenspec: 8.779580143694401e-05 
            9.058591846403974e-05 (desired)
   entspec: 5.519416835379740e-07 
            5.703438193477885e-07 (desired)
   APEflux: 8.986470103263862e-05 
            9.192940039964286e-05 (desired)
    KEflux: 1.768313116056767e-04 
            1.702621259427053e-04 (desired)
```

crocha700 replied:
Same here with ```tavestart=0 ```, but my results are slightly different than @rabernat's/

```   
    EKEdiss: 7.502521767193113e-11 
            6.368668363708053e-08 (desired)
    APEgen: 8.570574054449991e-08 
            8.842056320175081e-08 (desired)
APEgenspec: 8.778646755959426e-05 
            9.058591846403974e-05 (desired)
   entspec: 5.519110737187770e-07 
            5.703438193477885e-07 (desired)
   APEflux: 8.986199885636697e-05 
            9.192940039964286e-05 (desired)
    KEflux: 1.769198042761687e-04 
            1.702621259427053e-04 (desired)
```

rabernat replied:
Regarding the magnitude of ``EKEdiss``:
```python
In [2]: 7.502521767193113e-11 * (32**2)
Out[2]: 7.682582289605747e-08
```
Which is now also in a similar ballpark.

MFJansen replied:
@rabernat Good call on the tavestart! 
Since the model runs to right after the breaking of the unstable waves, it seems to me that the results are pretty sensitive to minor changes. It looks like machine-dependence probably can't explain the remaining differences, but a slight change in the numerics or the filter probably could.

MFJansen replied:
Regarding the normalization. The new normalization of EKEdiss seems to be more right, as can be inferred by comparing to rek*EKE2, which should give the same answer save for an O(1) factor given by the fractional depth of the lower layer. The latter actually appears to be missing in code though. So I think there still seem to be some issues with the diagnostics and how layer depths are accounted for (originally the diagnostics assumed two layers of equal depth. I thought I generalized this at some point, but apparently I missed some things...) I will try to review the energy budget diagnostics sometime soon. There is at least an easy consistency check that can be made, which is to check if the budget closes - I thought this was the case for the spectral energy budget, but maybe not for the bulk energy budget...

francispoulin replied:
To make things more concrete, is there a section in Geoff's book that we can cite and refer to how this should be calculated for the layered equations?  That might be useful for us and for other users as well, who want to know what convention we are following.

MFJansen replied:
I'm confused. From my reading of the code (whether right or not), these two should be identical:

In  [188]: m.get_diagnostic('EKEdiss')
Out[188]: 1.1141103759804538e-10

In  [189]: 2*m.rek*m.get_diagnostic('EKE')[-1]
Out[189]: 1.805314588712134e-10

But somehow they are not. Does anybody else see what I am missing?

rabernat replied:
from model.py
```python
        self.add_diagnostic('EKEdiss',
            description='total energy dissipation by bottom drag',
            function= (lambda self:
                       (self.rek*self.wv2*
                        np.abs(self.ph[-1])**2./(self.M**2)).sum())
        ) 
```

MFJansen replied:
@rabernat are you just providing some background information for others, or was that meant to be a response to my question? (If so I don't get it)

rabernat replied:
It was in response to your question. That is how the ``EKEdiss`` diagnostic is calculated. EKE instead is
```python
        self.add_diagnostic('EKE',
            description='mean eddy kinetic energy',
            function= (lambda self: 0.5*(self.v**2 + self.u**2).mean(axis=-1).mean(axis=-1))
        )
```

Just trying to help debug...

MFJansen replied:
Sorry, yes I saw that - I should have probably copied the relevant code segments in my comment in the first place. 
So why do these differ. It looks to me like they should be the same (safe for the factor 2*rek).

It looks like I can boil the issue down further. From just the last state of the model I get:

    (np.abs(m.p[-1])**2).mean()
    Out[221]: 1045198.6974801013
    
    (np.abs(m.ph[-1])**2./(m.M**2)).sum()
    Out[222]: 845712.10049875535

Shouldn't these be the same? (Parseval's identity)

crocha700 replied:
Gotta be careful with the Parseval's identity because we're using "real" FFT so we just have half of the coefficients. (That's why I created the "spec_var" method)

Here

```{python}
In [10]: (np.abs(m.p[-1])**2).mean()
Out[10]: 1038187.4151566233

In [11]: (np.abs(m.ph[-1])**2./(m.M**2)).sum()
Out[11]: 837904.39974625339
```
With ```spec_var```
```{python}
In [19]: m.spec_var(m.ph[-1])
Out[19]: 1038187.4151566236
```

Also, notice that the relative difference between our results is about 1%.

MFJansen replied:
Thanks @crocha700 ! I must admit I don't quite see the factor of two in the difference, but your method, spec_var seems to do get it right.

However, that means that the EKEdiss diagnostic in the model is definitely broken! And probably this affects other diagnostics too.

crocha700 replied:
It is not only a factor of two...notice that the real fft for ```Nx=Ny``` even has the shape ```(Nx/2+1,Ny)``.

MFJansen replied:
yeah, I got it (see my edit above)

MFJansen replied:
I will go through the diagnostics and see / fix where this causes an obvious problem.

crocha700 replied:
Yes. I think much of this is related to the fact that we're now using real fft...

MFJansen replied:
I need some help here: To compute the frictional dissipation correctly, I need to know the depth of the lowest layer - ideally as a fraction of the total depth, if we want to keep our energy budget in terms of volume average energies. Right now EKEdiss is a "core diagnostic", so it would work for all model types. If I see it correctly the way layer depths are handled differs between the two-layer and general layered model though, right? If I see it correctly I would have to use the vector self.Hi for the n-layer model, but this isn't defined for the two-layer model, right?

crocha700 replied:
```{python}
In [6]: m.H2/m.H
Out[6]: 0.8
```

or

```{python}
In [9]: 1./(1+m.delta)
Out[9]: 0.8
```

MFJansen replied:
Sorry, I guess I was unclear. I understand how to compute the layer depth ratio in the two layer model. 
The problem is that the EKEdiss diagnostic is implemented as a "core diagnostic", so the computation needs to work for all models (with the same code). I agree that it would be best to define Hi for the two layer model. It's not immediately obvious to me where this needs to be done (I'm not that familiar with the code anymore), but will have a look.

crocha700 replied:
yes I misunderstood you...so you can add something like

```
self.Hi = np.array([self.H1,self.H2])
```
just below [this](https://github.com/pyqg/pyqg/blob/master/pyqg/qg_model.py#L118)

MFJansen replied:
Yeah, I actually just started editing it with the goal of getting rid of self.H1 and self.H2 completely. It's easy to do in qg_model.py; I'm just unsure if self.H1 and self.H2 are used somewhere else?

crocha700 replied:
apparently only [here](https://github.com/pyqg/pyqg/blob/master/pyqg/qg_model.py#L217) and [here](https://github.com/pyqg/pyqg/blob/master/pyqg/qg_model.py#L226)

MFJansen replied:
Yes I found those. What I was unsure about is if they are used somewhere outside of qg_model.py. I don't see anything obvious...

crocha700 replied:
It shouldn't been used elsewhere outside this model subclass.

francispoulin replied:
Sorry for being a bit lost in the discussions here.  It seems that the code that you are looking at is the legacy code that computes the two-layer diagnostics.  I know that @crocha700 computed an N-layer version. Does that have diagnostics as well? 

Has anyone tried computing the fluxes with the two approaches and the diagnostics for a given snapshot to see if they yield the same result?  Maybe the n-layer isn't quite that developed.  I haven't managed to figure that much out yet.

crocha700 replied:
The layered model does compute diagnostics as documented [here](http://pyqg.readthedocs.org/en/latest/equations/notation_layered.html#energy-spectrum).

MFJansen replied:
The energy dissipation rate is a core diagnostic, so it's computed for any subclass model. 

The "APE generation rate" (really conversion from background flow to eddies) instead is so far only coded up as an additional diagnostic that exists in the two-layer sub-class.
@crocha700: you haven't coded up energy conversion rates for the n-layer model, right?

crocha700 replied:
[Here are the diagnostics](https://github.com/pyqg/pyqg/blob/master/pyqg/layered_model.py#L285) specific to the layered model subclass.

MFJansen replied:
Nice! @francispoulin is right, we should be able to compare results from a strict 2-layer setup to results from the layered model with an identical two layer configuration. That might be a good sanity check.

MFJansen replied:
I just pushed a minor fix, fixing the EKEdiss and APEgen diagnostics. 
I didn't do anything to the spectral diagnostics for now, since it seems less obvious what the "right" normalization here is.

MFJansen replied:
Save for a factor of nx*ny, the results for APEgen and EKEdiss now appear to be very close (within less than 1%) of the values that were commented out as "old" in the test. Here's what I get:
     m.get_diagnostic('APEgen')*32**2
    Out[35]: 2.522085319118987e-07
    
    m.get_diagnostic('EKEdiss')*32**2
    Out[36]: 1.4800760866049312e-07

Notice that the summed up spectra are not expected to agree due to the fft-rfft switch. Maybe that's not the best thing to look at anyway. In particular not the integrated spectral flux convergences - which should just integrate to zero anyway, when done right.

francispoulin replied:
That is a big step in the right direction!  Is that using the old 2-layer diagnostics or the new n-layer diagnostics?  Or maybe they are equal?

MFJansen replied:
@francispoulin: EKEdiss is a "core diagnostic" (in model.py), so the same code is used for the two layer and n-layer model. APEgen on the other hand is specific to the two-layer subclass. However, since @crocha700 has already implemented the APE generation spectrum in the n-layer subclass, it should be easy to compute the total APE generation there as well (by "appropriately" summing over all wavenumber - which isn't quite as trivial anymore with the rfft format of the spectra, but it's not that hard either). One should also be able to move the computation of the APE generation into the "core", though that may require a little bit more work.

francispoulin replied:
Thanks @MFJansen for spelling that out.  That helps.

MFJansen replied:
ps: I just realized I broke the barotropic model (which doesn't have Hi) - I'm on it...

crocha700 replied:
@MFJansen next time please try to branch off of master, and then submit the fixes as a pull request (PR). This procedure is described in the [development workflow](http://pyqg.readthedocs.org/en/stable/development.html)  that @rabernat put together. This approach helps us keep track of what's changed, and avoid code breaking. Every PR is tests using [ci-travis](https://travis-ci.org). @rabernat put a lot of effort to make out editions and inclusion of new features as smooth as possible. Let's try to follow that workflow.

crocha700 replied:
@MFJansen Thanks for fixing the energy dissipation ad APEgen diagnostics. I think things look better now.  Here I get with the 2-layer subclass:

```{python}
In [43]: m.get_diagnostic('EKEdiss')*(32**2)
Out[43]: 1.4809381615971648e-07
```

MFJansen replied:
oops, sorry, I'm not up to date with the procedures.

I guess differences \lessim 1% are to be expected for this test due to variations between machines. I think we could add a check of EKEdiss and APEgen back into test_twolayer_qg.

crocha700 replied:
Yes. Just updated my comment above. Even in a single machine there is differences between python distributions/ numpy versions, etc.

crocha700 replied:
Also, using the ```LayeredModel``` subclass with nz=2 and same config, I get, with anaconda,

```{python}
In [14]: m.get_diagnostic('EKEdiss')*(32**2)
Out[14]: 1.4808828678559061e-07

In [4]: m.get_diagnostic('EKEdiss')*(32**2)
Out[4]: 1.4802470150176962e-07

In [5]: APEgenspec = m.get_diagnostic('APEgenspec')
In [6]: APEgen = 2*APEgenspec
In [7]: APEgen[...,0] = APEgen[...,0]/2.
In [8]: APEgen[...,-1] = APEgen[...,-1]/2.
In [9]: (APEgen.sum()/m.M**2) * (32**2)
Out[9]: 2.5224036997131185e-07
```

crocha700 replied:
The differences are actually less than 0.07%, which are indeed very small.

crocha700 commented:
and with canopy and numpy 1.9.2

```{python} 
  EKEdiss: 7.507421618592886e-11 
            6.368668363708053e-08 (desired)
    APEgen: 8.577384674904105e-08 
            8.842056320175081e-08 (desired)
APEgenspec: 8.784672251018768e-05 
            9.058591846403974e-05 (desired)
   entspec: 5.522155131808690e-07 
            5.703438193477885e-07 (desired)
   APEflux: 8.987687995232196e-05 
            9.192940039964286e-05 (desired)
    KEflux: 1.765612305511634e-04 
            1.702621259427053e-04 (desired)
```

francispoulin commented:
Looks like @landscape-bot is getting agreement between two numpy's to 3 or so significant digits.    Not great but at least within the right order of magnitude. 

If we wanted to verify whether we should still consider the "desired" value, I wonder if we could compare this with another model that someone has lying around?  If this model has been tested by others then we should be able to figure out the order of magnitude rather easily. 

Maybe this is too much work though.

crocha700 commented:
I just restored the old content of ```test_two_layer_qg``` into ```test_reference_solns```. Thus, there is no change as far as that test is concerned. The goal of this PR is to simply organize this into a single tester and make the ```sqg``` and ```bt``` test faster.

We think should pick up this important discussion of ```test_two_ayer``` and other tests in different issues and PRs.

rabernat commented:
> Looks like @landscape-bot is getting agreement between two numpy's to 3 or so significant digits.

What do you mean by this? lanscape doesn't run any tests. Do you mean travis-ci? If so, this test suite does not actually include the diagnostic checks we are debating.

rabernat commented on pyqg/tests/test_reference_solns.py:
```diff
@@ -0,0 +1,244 @@
+        diagnostic_results = {
+                'APEgen': 2.5225558013107688e-07  / (m.nx**2),
+                'EKEdiss': 1.4806764171539711e-07 / (m.nx**2),           
+                }
```
Wow, lots of stuff happened last night!

I see that ``APEgen`` and ``EKEdiss`` have now made it into the test. This is definitely an improvement.

There are now lots of commented out / skipped diagnostic tests below. Maybe in the comments we should just clarify what they are, where they came from, and whether we plan to fix them or drop them. For our own sanity in the future.

crocha700 commented:
I managed to make all tests pass against Malte's original numbers to within less 1% (most tests actually pass within 0.1%). Towards that end, I implemented a [new diagnostics tool that performs the "spectral sum"](http://git.io/vlNqu) taking into consideration that we're using real FFT. We have to use that as oppose to .sum() when calculating the total sum of a spectral variable in order to satisfy Parseval's theorem (this is similar to `spec_var`, and in fact `spec_var` could leverage on `spec_sum` but that will  wait for the next PR). Other necessary hacks were to compare EKE = EKE1+EKE2 instead of individual components, and some normalization problems.

The normalization issue is not settled, but please let's move that discussion back to #90.

rabernat commented:
I think this is ready to merge (once landscape does it's thing).

crocha700 commented:
Just to summarize we identified three problems:

1. The ```tavestart``` bug
 That lead to the change to new values (now everything is restored to Malte's original values, a.k.a old values; I restored all old values in f7d8ff3; this PR)
2. `EKEdiss` and `APEgen` were not being correctly calculated (a factor of H2/H) was missing (@MFJansen fixed that in ddd5559)
3. The sum over wavenumbers was not updated when we implemented real FFT (I fixed that in f7d8ff3; this PR).

MFJansen commented:
Just to clarify:
APEgen was not correctly because of 3 (since it was computed as a spectral sum) - the layer depth ratio was actually accounted for correctly. EKEdiss was wrong for two reasons: one being the spectral sum issue, and the other one being the layer depth issue.

MFJansen commented:
I'm glad we sorted this out!

I also agree that the normalization for the spectral budget terms still needs to be looked at, so let's keep issue #90 open. I'll try to look into this sometime soon, but definitely won't get to it this week.

crocha700 commented:
Great! I just updated my comment above so that we have a good summary of the changes.
